driver/src/main/java/oracle/nosql/driver/query/SortIter.java

```diff
@@ -25,35 +26,94 @@
  */
 public class SortIter extends PlanIter {
 
+    private class CompareFunction implements Comparator<MapValue> {
+
+        RuntimeControlBlock theRCB;
+
+        CompareFunction(RuntimeControlBlock rcb) {
+            theRCB = rcb;
+        }
+
+        @Override
+        public int compare(MapValue v1, MapValue v2) {
+
+            return Compare.sortResults(theRCB,
+                                       v1,
+                                       v2,
+                                       theSortFields,
+                                       theSortSpecs);
+        }
+    }
+
+    private static class ReverseCompareFunction implements Comparator<MapValue> {
+
+        CompareFunction theComparator;
+
+        ReverseCompareFunction(CompareFunction comparator) {
+            theComparator = comparator;
+        }
+
+        @Override
+        public int compare(MapValue v1, MapValue v2) {
+            return -theComparator.compare(v1, v2);
+        }
+    }
+
     private static class SortIterState extends PlanIterState {
 
-        final ArrayList<MapValue> theResults;
+        CompareFunction theComparator;
+
+        int theLimit = -1;
+
+        ArrayList<MapValue> theResults;
```

Comment:
Please add some comments here that explain why there are three collections (list, queue, array) and when/how each are used.

Reply:
Done.